src/include/OSL/wide.h

```diff
@@ -1335,565 +1391,1553 @@ struct alignas(VecReg<WidthT>) Block<Dual2<Color3>, WidthT>
+    Mask<WidthT> m_mask;
+};
+
+template<typename ElementT, int WidthT> struct MaskedImpl<ElementT[], WidthT> {
+    static constexpr int width = WidthT;
+    typedef ElementT ElementType;
+
+    explicit OSL_FORCEINLINE MaskedImpl(void* ptr_wide_data, int array_length,
+                                        Mask<WidthT> mask, int derivIndex)
+        : m_array_of_wide_data(assume_aligned(&block_cast<ElementType, WidthT>(
+            ptr_wide_data)[array_length * derivIndex]))
+        , m_array_length(array_length)
+        , m_mask(mask)
+    {
+    }
+
+    explicit OSL_FORCEINLINE MaskedImpl(const MaskedData<WidthT>& md,
+                                        int derivIndex)
+        : MaskedImpl(md.ptr(), md.type().arraylen, md.mask(), derivIndex)
+    {
+    }
+
+protected:
+    OSL_FORCEINLINE
+    MaskedImpl(const MaskedImpl& other, Mask<WidthT> mask)
+        : m_array_of_wide_data(other.m_array_of_wide_data)
+        , m_array_length(other.m_array_length)
+        , m_mask(mask)
+    {
+    }
+
+    friend struct Masked<ElementT[], WidthT>;
+
+    static OSL_FORCEINLINE bool supports(const MaskedData<WidthT>& md)
+    {
+        // NOTE: using bitwise & to avoid branches
+        return (md.type().arraylen != 0)
+               & WideTraits<ElementType>::matches(md.type());
+    }
+
+public:
+    // Must provide user defined copy constructor to
+    // get compiler to be able to follow individual
+    // data members through back to original object
+    // when fully inlined the proxy should disappear
+    OSL_FORCEINLINE
+    MaskedImpl(const MaskedImpl& other) noexcept
+        : m_array_of_wide_data(other.m_array_of_wide_data)
+        , m_array_length(other.m_array_length)
+        , m_mask(other.m_mask)
+    {
+    }
+
+    OSL_FORCEINLINE int length() const { return m_array_length; }
+
+    typedef MaskedUnboundedArrayLaneProxy<ElementType, WidthT> Proxy;
+
+    OSL_FORCEINLINE Proxy const operator[](int lane) const
+    {
+        return Proxy(m_array_of_wide_data, m_array_length, m_mask, lane);
+    }
+
+    OSL_FORCEINLINE Masked<ElementType, WidthT>
+    get_element(int array_index) const
+    {
+        OSL_DASSERT(array_index < m_array_length);
+        return Masked<ElementType, WidthT>(m_array_of_wide_data[array_index],
+                                           m_mask);
+    }
+
+    // implicit conversion to constant read only access
+    operator Wide<const ElementT[], WidthT>() const
+    {
+        return Wide<const ElementT[], WidthT>(m_array_of_wide_data,
+                                              m_array_length);
+    }
+
+private:
+    mutable Block<ElementType, WidthT>* m_array_of_wide_data;
+    int m_array_length;
+    Mask<WidthT> m_mask;
+};
+
+
+}  // namespace pvt
+
+template<typename DataT, int WidthT>
+struct Masked : public pvt::MaskedImpl<DataT, WidthT> {
+    static_assert(std::is_const<DataT>::value == false,
+                  "Masked<> is only valid when DataT is NOT const");
+
```

Comment:
I think that for this and other templates, maybe the overview of what they do and why should be here in doxygen-like comments? I know I saw you explain it elsewhere, but I think probably each of these classes should have their primary documentation right at their declaration.

Reply:
At the top there is a "overview" with documentation by each class's forward declaration, I could repeat (maybe more detailed) the specific portion above each class, or move it down above each class..

Reply:
Yeah, I think maybe just move the explanations at the top to be near each class so that if we want to generate automatic docs for them later, it will all just work. I think people mostly expect to find the main documentation for a class at the point of its declaration.

This isn't really critical, we can move it around later, but I'm just wanting to be very careful to leave all the right breadcrumbs for whoever needs to modify or use these classes next.